### PR TITLE
fix: https://github.com/magma/security/issues/140

### PR DESCRIPTION
**Pull request by lucasgonze**
bug: https://github.com/magma/security/issues/140

If ORDERED_SUPPORTED_INTEGRITY_ALGORITHM_LIST is empty, throw an error instead of inserting EIA0.

If ORDERED_SUPPORTED_INTEGRITY_ALGORITHM_LIST contains an invalid setting, throw an error instead of replacing with EIA0.

If ORDERED_SUPPORTED_INTEGRITY_ALGORITHM_LIST contains < 8 items, pad by repeating the last-place item rather than assuming EIA0.

Throw error if ORDERED_SUPPORTED_INTEGRITY_ALGORITHM_LIST contains an unrecognized setting.

Signed-off-by: Lucas Gonze <lucas@gonze.com>

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp
Comment (lucasgonze) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp]: Fixed

File: lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp
Comment (sebathomas) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp]: Isn't this case already handled by the `AssertFatal` above?

Comment (lucasgonze) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp]: Thank you.

// This check was from before I added the AssertFatal. 

This code is now:
```
          char default_integrity_algorithm = config_pP->nas_config.prefered_integrity_algorithm[num-1];
           for (i = num; i < 8; i++) {
             config_pP->nas_config.prefered_integrity_algorithm[i] = default_integrity_algorithm;
           }
```

**Comment by lucasgonze** [comment]
Updates so far from review process:
- Added clang-format
- Rewrote if ( num==0) block per @sebathomas 
- git rebase
- CC'd @ssanadhya

**Comment by lucasgonze** [comment]
These three failures don't seem compelling:
image

**Comment by lucasgonze** [comment]
I have marked this ready for review because all of the meaningful syntax-level checks are passing, but I think architectural impacts still need discussion.

File: lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp:1528]: Hi @lucasgonze, Instead of filling the last configured value from the list. Can we set to invalid value say "0b111" 
I think the similar changes are required for encryption/ciphering algorithm, at https://github.com/magma/magma/pull/14468/files#diff-108a949404fe1a919b6e0e91872fcdf52cbfb1c47fba2066c1a5267734ca7e61R1558

**Comment by sebathomas** [comment]
> These three failures don't seem compelling: image

Yes, codecov sometimes flags small decreases in code coverage incorrectly, you can ignore these checks.

File: lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp
Comment (lucasgonze) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp:1528]: I can make that change. How confident are you that a flag value like 0b111 won't create bugs? 

I feel like I should skim all the code that references prefered_integrity_algorithm.

Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp:1528]: I was under impression that '0b111' would be invalid value, but referring to spec indicated that this value maps to NAS_SECURITY_ALGORITHMS_EEA7. Thanks for instruct me to rethink on this.
Can we configure last configured value from the list, other than EIA0_ALG_ID. Because we want to avoid configuring "null encryption"

Comment (lucasgonze) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp:1528]: Good teamwork.

I am not sure I am understanding your point correctly. The code as it stands is using the last configured value from the list, rather than defaulting to EIA0_ALG_ID.

Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp:1528]: If integrity is configured with, ORDERED_SUPPORTED_INTEGRITY_ALGORITHM_LIST = [ "EIA2" , "EIA1", "EIA0" ]; then remaining indices of "config_pP->nas_config.prefered_integrity_algorithm" would be filled with "EIA0_ALG_ID". 
We will set the remaining indices with last configured but other than "EIA0_ALG_ID".
Considering above configuration, the last configured value is "EIA0",instead of this we can set  "EIA1".
Correct me if I am wrong.

Comment (lucasgonze) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp:1528]: So the code would be something like

`        ORDERED_SUPPORTED_INTEGRITY_ALGORITHM_LIST = [ "EIA2" , "EIA0" , "EIA1" ];`

instead of the current:

`        ORDERED_SUPPORTED_INTEGRITY_ALGORITHM_LIST = [ "EIA2" , "EIA1" , "EIA0" ];`

Comment (lucasgonze) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp:1528]: I'm checking out the impact in the 3gpp specs. Here's the document:
https://www.etsi.org/deliver/etsi_ts/133400_133499/133401/16.03.00_60/ts_133401v160300p.pdf

The section is 7.2.4.3 "Procedures for NAS algorithm selection".

Comment (rsarwad) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_config.cpp:1528]: Okay got it. As per given reference above.  The MME selects the
NAS algorithms which have the highest priority according to the ordered lists.
So in our code, the remaining indices needs to be filled with last configured value. I agree with whatever changes you have done is correct.
Thanks for providing the reference.

**Review by rsarwad — APPROVED** [review]
LGTM

**Comment by lucasgonze** [comment]
@rsarwad @sebathomas I think this PR is waiting on a review from @ulaskozat, but I suspect he is no longer actively contributing. He has no Github activity since August or Slack activity since September.

**Comment by sebathomas** [comment]
The reviewer overview is a bit confusing but this PR is ready to merge. Do you want it to be merged or to continue working on it @lucasgonze?

**Comment by lucasgonze** [comment]
I'm ready to merge it, @sebathomas .

**Comment by sebathomas** [comment]
Done. :+1:

**Comment by lucasgonze** [comment]
Fix verified by original reporter.

**Pull request merged**
